source-oracle/source_oracle/api.py

```diff
@@ -0,0 +1,265 @@
+) -> list[OracleColumn]:
+    async with pool.acquire() as conn:
+        with conn.cursor() as c:
+            sql_columns = ','.join(["t." + f.alias for (k, f) in OracleColumn.model_fields.items() if f.alias != 'COL_IS_PK' and not f.exclude])
+
+            query = f"""
+            SELECT {sql_columns}, NVL2(c.constraint_type, 1, 0) as COL_IS_PK FROM user_tab_columns t
+                LEFT JOIN (
+                        SELECT c.table_name, c.constraint_type, ac.column_name FROM all_constraints c
+                            INNER JOIN all_cons_columns ac ON (
+                                c.constraint_name = ac.constraint_name
+                                AND c.table_name = ac.table_name
+                                AND c.constraint_type = 'P'
+                            )
+                        ) c
+                ON (t.table_name = c.table_name AND t.column_name = c.column_name)
+            """
+
+            columns = []
+            await c.execute(query)
+            async for values in c:
+                cols = [col[0] for col in c.description]
+                row = dict(zip(cols, values))
+                row = {k: v for (k, v) in row.items() if v is not None}
+                columns.append(OracleColumn(**row))
+
+            return columns
+
+
+async def fetch_page(
+    # Closed over via functools.partial:
+    table: Table,
+    pool: oracledb.AsyncConnectionPool,
+    # Remainder is common.FetchPageFn:
+    log: Logger,
+    page: str | None,
+    cutoff: Tuple[str],
+) -> AsyncGenerator[Document | str, None]:
+    is_first_query = False
+    if page is None:
+        is_first_query = True
+        async with pool.acquire() as conn:
+            with conn.cursor() as c:
+                await c.execute(f"select min(ROWID) from {table.table_name}")
+                page = (await c.fetchone())[0]
+
+    query = template_env.get_template("backfill").render(table=table, rowid=page, max_rowid=cutoff[0], is_first_query=is_first_query)
+
+    log.info(query, page)
+
+    last_rowid = None
+    async with pool.acquire() as conn:
+        with conn.cursor() as c:
+            await c.execute(query)
+            async for values in c:
+                cols = [col[0] for col in c.description]
+                row = dict(zip(cols, values))
+                row = {k: v for (k, v) in row.items() if v is not None}
+                last_rowid = row[rowid_column_name]
+
+                doc = Document()
+                source = Document.Meta.Source(
+                    table=table.table_name,
+                    row_id=row[rowid_column_name]
+                )
+                doc.meta_ = Document.Meta(op='c', source=source)
+                for (k, v) in row.items():
+                    if k == rowid_column_name:
+                        continue
+                    setattr(doc, k, v)
+
+                yield doc
+
+    if last_rowid is not None:
+        yield last_rowid
```

Comment:
Currently I'm emitting this checkpoint at the end, but I think we can emit this every N rows to make sure we don't lose a lot of progress in case we fail mid-way through a backfill